### PR TITLE
[Merged by Bors] - feat: port Data.Fin.Tuple.Basic

### PR DESCRIPTION
**Pull request by maxwell-thum**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

Should maybe wait for

- [x] leanprover-community/mathlib#10346

### CODE REVIEW

**Comment by jcommelin** [comment]
Why is there "attempt 2" in the PR title? It shouldn't stay there when we merge this.

**Comment by eric-wieser** [comment]
Reiterating the message from the porting wiki, I would prefer that we get #10346 in first, as this change the file and the things it depends on a fair bit.

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,898 @@
...
+
+theorem tail_def {n : ℕ} {α : Fin (n + 1) → Type _} {q : ∀ i, α i} :
+    (tail fun k : Fin (n + 1) ↦ q k) = fun k : Fin n ↦ q k.succ :=
+  rfl
+#align fin.tail_def Fin.tail_def
+
+-- Porting note: I made this noncomputable because Lean seemed to think it should be
+/-- Adding an element at the beginning of an `n`-tuple, to get an `n+1`-tuple. -/
+noncomputable def cons (x : α 0) (p : ∀ i : Fin n, α i.succ) : ∀ i, α i := fun j ↦ Fin.cases x p j
+#align fin.cons Fin.cons
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Tuple/Basic.lean:72]: Is there a lean issue tracking this? It seems bizarre that `SomeInductive.cases` is noncomputable when the equation compiler is happy to produce something computable.

**Comment by maxwell-thum** [comment]
> Why is there "attempt 2" in the PR title? It shouldn't stay there when we merge this.

When I first tried porting this (https://github.com/leanprover-community/mathlib4/pull/1515) I got a strange error where `ToAdditive.lean` wouldn't build. I couldn't understand why (and I'm not super familiar with GitHub), so I deleted the branch and started over, running `start_port.sh` by hand. I'll change the PR titles.

**Comment by maxwell-thum** [comment]
> Reiterating the message from the porting wiki, I would prefer that we get #10346 in first, as this change the file and the things it depends on a fair bit.

My bad; I didn't see that message.

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,878 @@
...
+
+! This file was ported from Lean 3 source module data.fin.tuple.basic
+! leanprover-community/mathlib commit 7c523cb78f4153682c2929e3006c863bfef463d0
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Fin.Basic
+import Mathlib.Data.Pi.Lex
+import Mathlib.Data.Set.Intervals.Basic
+import Mathlib.Tactic.LibrarySearch -- **TODO** delete this later
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Fin/Tuple/Basic.lean:14]:
```suggestion
```

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,878 @@
...
+* `Fin.snoc` : adding an element at the end of an `n`-tuple, to get an `n+1`-tuple. The name `snoc`
+  comes from `cons` (i.e., adding an element to the left of a tuple) read in reverse order.
+* `Fin.insertNth` : insert an element to a tuple at a given position.
+* `Fin.find p` : returns the first index `n` where `p n` is satisfied, and `none` if it is never
+  satisfied.
+
+-/
+
+set_option autoImplicit false -- **TODO** delete this later
+
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Fin/Tuple/Basic.lean:38]:
```suggestion
```

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,878 @@
...
+#align fin.cons_succ Fin.cons_succ
+
+@[simp]
+theorem cons_zero : cons x p 0 = x := by simp [cons]
+#align fin.cons_zero Fin.cons_zero
+
+/-- Updating a tuple and adding an element at the beginning commute. -/
+@[simp]
+theorem cons_update : cons x (update p i y) = update (cons x p) i.succ y :=
+  by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Fin/Tuple/Basic.lean:89]:
```suggestion
theorem cons_update : cons x (update p i y) = update (cons x p) i.succ y := by
```

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,878 @@
...
+but is expected to have type
+  forall {α : Type.{u1}} {P : forall {n : Nat}, ((Fin n) -> α) -> Sort.{u2}}, (P (OfNat.ofNat.{0} Nat 0 (OfNat.mk.{0} Nat 0 (Zero.zero.{0} Nat Nat.hasZero))) (Fin.elim0ₓ.{succ u1} (fun (ᾰ : Fin (OfNat.ofNat.{0} Nat 0 (OfNat.mk.{0} Nat 0 (Zero.zero.{0} Nat Nat.hasZero)))) => α))) -> (forall {n : Nat} (x₀ : α) (x : (Fin n) -> α), (P n x) -> (P (HAdd.hAdd.{0, 0, 0} Nat Nat Nat (instHAdd.{0} Nat Nat.hasAdd) n (OfNat.ofNat.{0} Nat 1 (OfNat.mk.{0} Nat 1 (One.one.{0} Nat Nat.hasOne)))) (Fin.cons.{u1} n (fun (ᾰ : Fin (HAdd.hAdd.{0, 0, 0} Nat Nat Nat (instHAdd.{0} Nat Nat.hasAdd) n (OfNat.ofNat.{0} Nat 1 (OfNat.mk.{0} Nat 1 (One.one.{0} Nat Nat.hasOne))))) => α) x₀ x))) -> (forall {n : Nat} (x : (Fin n) -> α), P n x)
+Case conversion may be inaccurate. Consider using '#align fin.cons_induction Fin.consInductionₓ'. -/
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Fin/Tuple/Basic.lean:171]:
```suggestion
```

I don't know when we need the tiny x, but the comment should go

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 6]
```diff
@@ -0,0 +1,878 @@
...
+#align fin.cons_cases Fin.consCases
+
+@[simp]
+theorem cons_cases_cons {P : (∀ i : Fin n.succ, α i) → Sort v} (h : ∀ x₀ x, P (Fin.cons x₀ x))
+    (x₀ : α 0) (x : ∀ i : Fin n, α i.succ) : @consCases _ _ _ h (cons x₀ x) = h x₀ x := by
+  rw [consCases, cast_eq]
+  congr
+#align fin.cons_cases_cons Fin.cons_cases_cons
+
+-- Porting note: I am not perfectly confident in my choice to add `h0 h` to `consInduction` at end
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Fin/Tuple/Basic.lean:165]: If you mean the extra arguments in the recursive call, that seems standard in lean4
```suggestion
```

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 7]
```diff
@@ -0,0 +1,878 @@
...
+theorem init_snoc : init (snoc p x) = p := by
+  ext i
+  have h' := Fin.cast_lt_castSucc i i.is_lt
+  simp [init, snoc, i.is_lt, h']
+  convert cast_eq rfl (p i)
+#align fin.init_snoc Fin.init_snoc
+
+@[simp]
+theorem snoc_cast_succ : snoc p x (castSucc i) = p i :=
+  by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Fin/Tuple/Basic.lean:355]:
```suggestion
theorem snoc_cast_succ : snoc p x (castSucc i) = p i := by
```

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 8]
```diff
@@ -0,0 +1,878 @@
...
+/-- Insert an element into a tuple at a given position. For `i = 0` see `Fin.cons`,
+for `i = Fin.last n` see `Fin.snoc`. See also `Fin.succAboveCases` for a version elaborated
+as an eliminator. -/
+def insertNth (i : Fin (n + 1)) (x : α i) (p : ∀ j : Fin n, α (i.succAbove j)) (j : Fin (n + 1)) :
+    α j :=
+  succAboveCases i x p j
+#align fin.insert_nth Fin.insertNth
+
+@[simp]
+theorem insert_nth_apply_same (i : Fin (n + 1)) (x : α i) (p : ∀ j, α (i.succAbove j)) :
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Fin/Tuple/Basic.lean:551]: Many lemmas about `insertNth` need to use that casing in their name

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 9]
```diff
@@ -0,0 +1,878 @@
...
+    insertNth i x p i = x := by simp [insertNth, succAboveCases]
+#align fin.insert_nth_apply_same Fin.insert_nth_apply_same
+
+@[simp]
+theorem insert_nth_apply_succ_above (i : Fin (n + 1)) (x : α i) (p : ∀ j, α (i.succAbove j))
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Fin/Tuple/Basic.lean:556]: Also `succAbove`

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 10]
```diff
@@ -0,0 +1,878 @@
...
+
+theorem le_insert_nth_iff {i : Fin (n + 1)} {x : α i} {p : ∀ j, α (i.succAbove j)} {q : ∀ j, α j} :
+    q ≤ i.insertNth x p ↔ q i ≤ x ∧ (fun j ↦ q (i.succAbove j)) ≤ p := by
+  simp [Pi.le_def, forall_iff_succ_above i]
+#align fin.le_insert_nth_iff Fin.le_insert_nth_iff
+
+open Set
+
+-- Porting note: Had to up maxHeartbeats
+set_option maxHeartbeats 1000000
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Fin/Tuple/Basic.lean:700]: Can we limit this to only the declaration that needs it?

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 11]
```diff
@@ -0,0 +1,878 @@
...
+      --exact find_spec _ (@_root_.cast ((find fun i ↦ p (castLt i (_ : ↑i < Nat.succ n))) = some j)
+       --((find fun (i : Fin <| n+0) => p (castLt i (_ : ↑i < Nat.succ (n+0)))) = some j) _ h)
+      sorry
+      rfl
+#align fin.find_spec Fin.find_spec
+
+#check Membership.mem
+
+/-- `find p` does not return `none` if and only if `p i` holds at some index `i`. -/
+theorem is_some_find_iff :
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Fin/Tuple/Basic.lean:769]: `isSome_find_iff`

**Review by Ruben-VandeVelde — CHANGES_REQUESTED** [review]

File: Mathlib/Data/Fin/Tuple/Basic.lean (hunk 10 above)
Comment (jcommelin) [reply Mathlib/Data/Fin/Tuple/Basic.lean]: No longer needed

File: Mathlib/Data/Fin/Tuple/Basic.lean (hunk 2 above)
Comment (jcommelin) [reply Mathlib/Data/Fin/Tuple/Basic.lean:14]: Done

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 12]
```diff
@@ -430,8 +427,7 @@ theorem update_snoc_last : update (snoc p x) (last n) z = snoc p z := by
 theorem snoc_init_self : snoc (init q) (q (last n)) = q := by
   ext j
   by_cases h : j.val < n
-  · have : j ≠ last n := ne_of_lt h
-    simp [h, update_noteq, this, snoc, init, castSucc_cast_lt]
+  · simp [h, update_noteq, snoc, init, castSucc_cast_lt]
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Fin/Tuple/Basic.lean:430]: Can you squeeze this simp while you're here?

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 13]
```diff
@@ -0,0 +1,839 @@
...
+
+/-- Updating an element and taking the beginning commute. -/
+@[simp]
+theorem init_update_cast_succ : init (update q (castSucc i) y) = update (init q) i y := by
+  ext j
+  by_cases h : j = i
+  · rw [h]
+    simp [init]
+  · simp [init, h]
+#align fin.init_update_cast_succ Fin.init_update_cast_succ
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Fin/Tuple/Basic.lean:453]:
```suggestion
theorem init_update_castSucc : init (update q (castSucc i) y) = update (init q) i y := by
  ext j
  by_cases h : j = i
  · rw [h]
    simp [init]
  · simp [init, h]
#align fin.init_update_cast_succ Fin.init_update_castSucc
```

**Review by Ruben-VandeVelde — APPROVED** [review]

**Comment by Ruben-VandeVelde** [comment]
> Reiterating the message from the porting wiki, I would prefer that we get [leanprover-community/mathlib#10346](https://github.com/leanprover-community/mathlib/pull/10346) in first, as this change the file and the things it depends on a fair bit.

For the record, we discussed this [on zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Data.2EFin.2ETuple.2EBasic.20mathlib4.231516) and ended up not waiting for this, because mathlib3port is not being updated right now. I'm happy to help re-port once it's back up (though you may need to remind me).

File: Mathlib/Data/Fin/Tuple/Basic.lean (hunk 1 above)
Comment (ChrisHughes24) [reply Mathlib/Data/Fin/Tuple/Basic.lean:69]: The compiler ignores the actual Lean term generated by pattern matching and just compiles the pattern matching directly. I'm sure it's on the todo list of the developers.

**Pull request closed without merging**